// market-csgo/lib/Utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bignumber_js_1 = tslib_1.__importDefault(require("bignumber.js"));
const EnumCurrency_1 = require("./declarations/enums/EnumCurrency");
class UtilsClass {
    IsStringMarketHashName(name) {
        return name.indexOf('_') < 1;
    }
    GetTradeObject(tradeUrl) {
        const tradeUrlRegex = /^https:\/\/steamcommunity.com\/tradeoffer\/new\/\?partner=([0-9]{6,32})&token=([a-zA-Z0-9]{3,12})$/i;
        if (typeof tradeUrl === 'string' && tradeUrlRegex.test(tradeUrl)) {
            const executed = tradeUrlRegex.exec(tradeUrl);
            if (executed !== null) {
                return {
                    Partner: executed[1],
                    Token: executed[2]
                };
            }
        }
        else if (tradeUrl !== null && typeof tradeUrl === 'object') {
            return tradeUrl;
        }
        return {
            Token: undefined,
            Partner: undefined
        };
    }
    ConvertPriceToCoins(price, currency) {
        if (currency === EnumCurrency_1.EnumCurrency.RUB) {
            return Math.ceil(new bignumber_js_1.default(price).multipliedBy(100).toNumber());
        }
        return Math.ceil(new bignumber_js_1.default(price).multipliedBy(1000).toNumber());
    }
}
exports.UtilsClass = UtilsClass;
exports.Utils = new UtilsClass();
