// market-csgo/lib/declarations/Endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Endpoints;
(function (Endpoints) {
    Endpoints["Prices"] = "prices/{CURRENCY_CODE}.json";
    Endpoints["PricesWithClassAndInstances"] = "prices/class_instance/{CURRENCY_CODE}.json";
    Endpoints["TradeRequestTake"] = "trade-request-take?key={SECRET_KEY}";
    Endpoints["TradeRequestTakeWithBotId"] = "trade-request-take?key={SECRET_KEY}&bot={BOT_ID}";
    Endpoints["TradeRequestGive"] = "trade-request-give?key={SECRET_KEY}";
    Endpoints["TradeRequestGiveP2P"] = "trade-request-give-p2p?key={SECRET_KEY}";
    Endpoints["TradeRequestGiveP2PAll"] = "trade-request-give-p2p-all?key={SECRET_KEY}";
    Endpoints["Ping"] = "ping?key={SECRET_KEY}";
    Endpoints["MyInventory"] = "my-inventory?key={SECRET_KEY}";
    Endpoints["AddToSale"] = "add-to-sale?key={SECRET_KEY}&id={ITEM_ID}&price={PRICE}&cur={CURRENCY_CODE}";
    Endpoints["SetPrice"] = "set-price?key={SECRET_KEY}&item_id={ITEM_ID}&price={PRICE}&cur={CURRENCY_CODE}";
    Endpoints["RemoveAllFromSale"] = "remove-all-from-sale?key={SECRET_KEY}";
    Endpoints["Items"] = "items?key={SECRET_KEY}";
    Endpoints["Trades"] = "trades?key={SECRET_KEY}";
    Endpoints["TradesExtended"] = "trades?key={SECRET_KEY}&extended=1";
    Endpoints["BuyWithMarketHashName"] = "buy?key={SECRET_KEY}&hash_name={ITEM_ID}&price={PRICE}&cur={CURRENCY_CODE}";
    Endpoints["BuyWithItemId"] = "buy?key={SECRET_KEY}&id={ITEM_ID}&price={PRICE}&cur={CURRENCY_CODE}";
    Endpoints["BuyWithMarketHashNameAndCustomId"] = "buy?key={SECRET_KEY}&hash_name={ITEM_ID}&price={PRICE}&cur={CURRENCY_CODE}&custom_id={CUSTOM_ID}";
    Endpoints["BuyWithItemIdAndCustomId"] = "buy?key={SECRET_KEY}&id={ITEM_ID}&price={PRICE}&cur={CURRENCY_CODE}&custom_id={CUSTOM_ID}";
    Endpoints["BuyForWithMarketHashName"] = "buy-for?key={SECRET_KEY}&hash_name={ITEM_ID}&price={PRICE}&cur={CURRENCY_CODE}&partner={PARTNER}&token={TOKEN}";
    Endpoints["BuyForWithItemId"] = "buy-for?key={SECRET_KEY}&id={ITEM_ID}&price={PRICE}&cur={CURRENCY_CODE}&partner={PARTNER}&token={TOKEN}";
    Endpoints["BuyForWithMarketHashNameAndCustomId"] = "buy-for?key={SECRET_KEY}&hash_name={ITEM_ID}&price={PRICE}&cur={CURRENCY_CODE}&partner={PARTNER}&token={TOKEN}&custom_id={CUSTOM_ID}";
    Endpoints["BuyForWithItemIdAndCustomId"] = "buy-for?key={SECRET_KEY}&id={ITEM_ID}&price={PRICE}&cur={CURRENCY_CODE}&partner={PARTNER}&token={TOKEN}&custom_id={CUSTOM_ID}";
    Endpoints["GetBuyInfoByCustomId"] = "get-buy-info-by-custom-id?key={SECRET_KEY}&custom_id={CUSTOM_ID}";
    Endpoints["GetListBuyInfoByCustomId"] = "get-list-buy-info-by-custom-id?key={SECRET_KEY}{CUSTOM_IDS}";
    Endpoints["History"] = "history?key={SECRET_KEY}&date={START_DATE}";
    Endpoints["HistoryWithEndDate"] = "history?key={SECRET_KEY}&date={START_DATE}&date_end={END_DATE}";
    Endpoints["GetMoney"] = "get-money?key={SECRET_KEY}&cur={CURRENCY_CODE}";
    Endpoints["GoOffline"] = "go-offline?key={SECRET_KEY}";
    Endpoints["UpdateInventory"] = "update-inventory?key={SECRET_KEY}";
    Endpoints["TransferDiscounts"] = "transfer-discounts?key={SECRET_KEY}&to={TO_SECRET_KEY}";
    Endpoints["GetMySteamId"] = "get-my-steam-id?key={SECRET_KEY}";
    Endpoints["SearchItemByHashName"] = "search-item-by-hash-name?key={SECRET_KEY}&hash_name={MARKET_HASH_NAME}&cur={CURRENCY_CODE}";
    Endpoints["SearchItemByHashNameSpesific"] = "search-item-by-hash-name-specific?key={SECRET_KEY}&hash_name={MARKET_HASH_NAME}&cur={CURRENCY_CODE}";
    Endpoints["SearchListItemsByHashNameAll"] = "search-list-items-by-hash-name-all?key={SECRET_KEY}&cur={CURRENCY_CODE}{MARKET_HASH_NAMES}";
    Endpoints["GetListItemsInfo"] = "get-list-items-info?key={SECRET_KEY}&cur={CURRENCY_CODE}{MARKET_HASH_NAMES}";
    Endpoints["Test"] = "test?key={SECRET_KEY}";
})(Endpoints = exports.Endpoints || (exports.Endpoints = {}));
